lib/fs_common.ml

```diff
@@ -57,27 +57,31 @@ let map_error err reqd_string =
                                     (reqd_string ^ ": error " ^ string_of_int i))
 
 let read_impl base name off reqd_len =
+  if reqd_len < 0 then return (`Error (`Unknown_error "can't read negative bytes"
+                                      ))
+  else begin
   let fullname = resolve_filename base name in
-  try_lwt
-    Lwt_unix.openfile fullname [Lwt_unix.O_RDONLY] 0 >>= fun fd ->
-    Lwt_unix.lseek fd off Lwt_unix.SEEK_SET >>= fun _seek -> (* very little we can do with _seek *)
-    let st =
-      Lwt_stream.from (fun () ->
-        let buf = Cstruct.create 4096 in
-        lwt len = Lwt_cstruct.read fd buf in
-        match len with
-        | 0 ->
-           Lwt_unix.close fd
-           >>= fun () -> return None
-        | len ->
-           return (Some (Cstruct.sub buf 0 (min len reqd_len)))
-      )
-    in
-    Lwt_stream.to_list st >>= fun bufs ->
-    match reqd_len with
-    | 0 -> return (`Ok [])
-    | n -> return (`Ok bufs)
-  with Unix.Unix_error (ex, _, _) -> return (map_error ex name)
+    try_lwt
```

Comment:
while you are it, can you remove camlp4 syntax extension? `Lwt.catch` works well in that case :-)